### PR TITLE
iOS: UTI - New Navigation, Header Buttons, Menu Items and No Toggle in Chat Mode

### PR DESCRIPTION
**Pull request by aataraxiaa**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1206488453854252/task/1214995978971487?focus=true
Tech Design URL:
CC:

### Description

Reworks Duck.ai chat navigation in the unified-toggle-input flow.

1. **Boundary preservation** — any user-initiated navigation crossing Duck.ai ↔ web from a non-empty tab opens a new tab instead of replacing the origin. Web↔web, chat↔chat, and NTP transformations stay in place. `AIBoundaryNavigationDecision` is wired through every entry path: omnibar submit, suggestions, bookmarks, history, deep links, in-content link taps in `TabViewController.decidePolicyFor`.

2. **New chat tab header** — top-left `[X, chat-history]` as separate pills, top-right `[+ menu, tab counter]` in one pill. X uses `closeTab(behavior: .onlyClose)` so selection follows the tab-switcher rule (previous tab / next when the chat was first; a fresh NTP only when the chat was the only tab). Plus menu has 5 rows: **New Chat** / **New Voice Chat** / — / **New Tab** (keyboard down, inherits persisted toggle) / **New Search** (keyboard up, forces `.search` for this entry without committing to storage) / **New Fire Tab**.

3. **UTI on Duck.ai tabs** — flanked voice button replaced by the app-menu hamburger. Search↔Duck.ai toggle hidden on Duck.ai tabs (`aiChatTabHideToggle`).

4. **App menu on Duck.ai tabs (F9 layout)** — header tiles `[New Chat | Chat History | Chat Settings]`; "Settings" becomes a standalone list row directly under the header. The old "Duck.ai Chats" middle-section row and the bottom "Duck.ai Settings" row are removed (both promoted to header tiles). Chat History uses the same glyph as the chat-history pill in the top chrome. Non-Duck.ai tab menus are unchanged.

One deliberate legacy (feature-off) behavior change: same-frame chat→web link taps always open a new tab, so model-emitted citations no longer destroy the chat.

12 unit tests in `AIBoundaryNavigationDecisionTests` cover the decision matrix. Coordinator gating tests cover the `aiChatTabHideToggle` truth table and app-menu delegate forwarding.

### Testing Steps

**Setup.** iOS Browser scheme. Two flags involved:
- `unifiedToggleInput` — master gate; all chrome and boundary changes are off when this is off.
- `aiChatTabHideToggle` — new failsafe (default ON, kill-switch via privacy config). When on, the Search↔Duck.ai toggle is hidden on Duck.ai tabs regardless of the user's "Show Duck.ai Toggle" setting.

#### Phase 1 — Migration setup (both flags OFF)
Build the state Phase 2 will migrate. Tap-through only — model-emitted web link clicks open in a separate tab in the legacy flow, so use the input-bar Search↔Duck.ai toggle to interleave history within a single tab.

- **S1a. Mixed-history tab, parked on Duck.ai.** In a new tab, submit a search like "weather" → toggle the input bar to Duck.ai and submit a prompt → toggle back to search and submit another query → toggle back to Duck.ai and submit one more prompt. **Leave the tab on the Duck.ai side.** This is the "stuck on Duck.ai with web entries earlier in history" migration case.
- **S1b. Mixed-history tab, parked on search.** Repeat the S1a sequence in another new tab, but end on the search side after the final toggle+submit. This is the "land on web, navigate back into a Duck.ai entry" migration case.
- **S2. Chat-only tab.** Open Duck.ai in a new tab, submit a few prompts. Stay on Duck.ai.
- **S3. Search-only tab.** Submit a web search, stay on the SERP.
- **S4. Apply flags + relaunch.** Background the app, enable both flags (`unifiedToggleInput`, `aiChatTabHideToggle`), **fully kill the app** (swipe up from app switcher), then launch it again. Flags are only read on app launch — foregrounding the existing process won't pick them up.

#### Phase 2 — Migration verification (flags now ON, after S4 relaunch)
Tests run against the specific tabs created in Phase 1, in their newly-migrated state.

- **M1a. S1a tab (mixed, parked on Duck.ai).** Switching to it lands on Duck.ai with the new chat header (X, chat-list pill, Plus + tab count). UTI shows no toggle. Bottom UTI has the hamburger (no voice). **Document the "stuck" UX:** the earlier web entries in this tab's history are not reachable from chat chrome (no back/forward exposed). X / Plus menu / chat-list are the only escape paths. X follows the tab-switcher selection rule.
- **M1b. S1b tab (mixed, parked on search).** Switching to it lands on a web page with the regular browser chrome. Use the toolbar back button to walk backward through this tab's history until you reach a Duck.ai entry. **Verify the chat chrome appears** on landing, and the same "stuck" condition applies — further back-traversal isn't exposed from the chat side. X closes the tab.
- **M2. S2 tab (chat-only).** New chat header. UTI shows no toggle. Bottom UTI hamburger (no voice). X closes the tab per the tab-switcher rule.
- **M3. S3 tab (search-only).** Regular browser chrome — no visible change.

#### Phase 3 — Boundary cross (new tab opens, origin survives)
- **B1.** Web → omnibar `duck.ai` or a prompt.
- **B2.** Web → recent chat
- **B3.** Web → search assist (search “Bread Recipe” and then submit followup prompt)
- **B4.** Chat → related search tap
- **B5.** Chat → bookmark

#### Phase 4 — In-place (no boundary cross)
- **P1.** NTP → submit anything → NTP transforms in place.
- **P2.** Web → web URL / search query → in place.
- **P3.** Chat → another `duck.ai` URL → in place.
- **P4.** Recent-chat selection from NTP → opens in place.

#### Phase 5 — Chat tab chrome
- **C1. Header layout.** Top-left: two separate circular pills `[X]` and `[chats sidebar]`. Top-right: one pill containing `[+, tab count]`. No hamburger in the header.
- **C2. Plus menu — 5 rows** with a divider after row 2:
    - New Chat
    - New Voice Chat
    - ─── divider ───
    - **New Tab** — keyboard *down*, NTP inherits the persisted toggle preference (no mode override).
    - **New Search** — keyboard *up*, forces `.search` for this entry only. Verify `toggleModeStorage` is NOT mutated (next organic activation restores the prior preference).
    - **New Fire Tab** — fire mode NTP (keyboard up, unchanged).
- **C3. Close (X) — tab-switcher selection rule.**
    - a. If multiple tabs closing duck.ai tab lands user on tab-1 (same as closing tab via tab switcher)
    - b. Duck.ai is the only tab → close → fresh NTP, counter "1" (normal mode). In fire mode the fire empty-state appears instead of an auto-inserted NTP.
- **C4. Bottom UTI on Duck.ai tab.** Fire button at left, single-line composer pill in the middle, hamburger app-menu at right (no voice waveform).
- **C5. No UTI toggle** Search↔Duck.ai toggle is hidden on Duck.ai tabs regardless of Settings → "Show Duck.ai Toggle".
- **C6. App menu on Duck.ai tab — F9 layout:**
    - Header tiles: `[New Chat | Chat History | Chat Settings]`.
    - "Settings" standalone row directly under the header tiles.
    - No "Duck.ai Chats" row in the middle section.
    - No "Duck.ai Settings" row at the bottom.
    - Chat History tile uses the same glyph as the chat-history pill in the top chrome.
    - Chat History opens the recent-chats sidebar (same destination as the pill).
    - Chat Settings opens the Duck.ai-specific settings screen.
- **C12. App menu on non-Duck.ai tab** — unchanged. Header `[New Tab | New Chat | Settings]`. "Duck.ai Chats" + "Duck.ai Settings" rows still present where they were.

#### Phase 6 — Failsafe flag (`unifiedToggleInput` ON, `aiChatTabHideToggle` OFF)
Flags are only read on app launch — flip `aiChatTabHideToggle` OFF, then fully kill the app and relaunch before testing (same as S4).
- **T1.** Duck.ai tab + user setting "Show Duck.ai Toggle" ON → toggle visible in the pill (failsafe rollback path).
- **T2.** Search tab + user setting ON → toggle visible (unaffected — only Duck.ai tabs are gated).

#### Phase 7 — Legacy regression (both flags OFF)
- **L1.** All entry points navigate in place (unchanged): omnibar / favorites / bookmarks / history / suggestions.
- **L2.** Chat tab uses legacy chrome — toolbar at the bottom, no `[X, chat-history, +, tab count]`.
- **L3. Deliberate change** — same-frame chat→web link tap opens a *new* tab (previously replaced the chat). Intentional — Duck.ai tab preservation treated as a universal property.

### Impact and Risks

**Medium** — touches every navigation entry point and Duck.ai tab chrome. Rollback paths exist via `unifiedToggleInput` (turn off all of it) and `aiChatTabHideToggle` (selectively re-show the toggle on Duck.ai tabs). Non-UTI builds are unchanged except for L3.

###### Internal references:
[Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f) | [Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552) | [Tech Design Template](https://app.asana.com/0/59792373528535/184709971311943)

---

> [!NOTE]
> **Medium Risk**
> Touches all major navigation entry points and Duck.ai chrome; behavior is gated by `unifiedToggleInput` and `aiChatTabHideToggle`, with one legacy change (chat→web links always open a new tab).
> 
> **Overview**
> Reworks **unified toggle input** Duck.ai tab navigation: a new **chat tab header** (close tab, recent chats, Plus menu with new chat/voice/tab/search/fire, tab switcher) replaces back/forward and the header app menu; the **flanked UTI** swaps voice for **hamburger → app menu**, and **`aiChatTabHideToggle`** hides the Search↔Duck.ai toggle on Duck.ai tabs (remote kill-switch, default on).
> 
> **`AIBoundaryNavigationDecision`** centralizes when Duck.ai ↔ web crossings open a **new tab** vs load in place; **`loadUrlRespectingAIBoundary`** wires that through omnibar, suggestions, bookmarks, history, chat history, and **`TabViewController`** same-frame link taps (chat→web always spawns a new tab). **`openAIChat`** / **`loadUrlInNewTab`** gain completion and prompt staging for new-tab chat opens.
> 
> **Duck.ai app menu** (UTI on): F9 header tiles (new chat, chat history, chat settings) and a chat-first list; browsing menu shows Duck.ai logo when UTI is available. Maestro and unit tests cover header, boundary matrix, toggle kill-switch, and tab removal edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 36ad6fefe1c83a98d977d134389a1dab10233791. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: iOS/DuckDuckGo/MainViewController.swift [hunk 1]
```diff
@@ -3717,7 +3781,10 @@ extension MainViewController: OmniBarDelegate {
 
     func onChatHistorySelected(url: URL) {
         postIdleSessionInstrumentation.sessionEnded(reason: .chatSelected)
-        loadUrlInNewTab(url, inheritedAttribution: nil)
+        // Route through the boundary helper so an NTP transforms in place into the chat tab,
+        // same as a prompt submission. Web/empty tab → chat URL crossing the boundary spawns a
+        // new tab; chat → chat stays in place. Matches `onPromptSubmitted` semantics.
+        loadUrlRespectingAIBoundary(url)
```
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift:3782]: Intentional. `loadUrlRespectingAIBoundary` applies the same boundary rule used everywhere else in this PR: NTP transforms in place, web→chat opens a new tab, chat→chat stays in place. Selecting a chat from the recent-chats list is semantically the same as submitting a new prompt — `onPromptSubmitted` already routes through the boundary helper, and the `+ → New Chat` button in the header also resets in place. The previous `loadUrlInNewTab` always-new-tab behavior would have been the odd one out. If you want to preserve the current chat, open a new tab first.

File: iOS/DuckDuckGo/MainViewController.swift [hunk 2]
```diff
@@ -1928,6 +1934,7 @@ class MainViewController: UIViewController {
             // Check if an existing tab with the same URL should be reused.
             else if reuseExisting != .none, let existing = tabManager.first(withUrl: url) {
                 selectTab(existing)
+                completion?()
```
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift:1939]: Documented intent — see the docstring above this function (lines 1918-1922). The URL-reuse branch hits `selectTab(existing)`, which already drives the chrome updates via the same path the user takes when picking a tab from the switcher; the `refreshOmniBar/refreshTabIcon/refreshControls` calls at the bottom of the function are needed for the new-tab / NTP-reuse paths but would be redundant here. The completion closure used by `openAIChat` only needs the target tab to be current (for `setPayload`), which `selectTab` guarantees — it does not depend on chrome state. The docstring explicitly states callers must not rely on that.

File: iOS/DuckDuckGo/MainViewController.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift:3782]: By design —  is intentional here so that selecting a chat from the sidebar while on an NTP transforms the NTP in place (matching `onPromptSubmitted` semantics), and chat → chat stays in place as the rest of the codebase enforces via `AIBoundaryNavigationDecision`. Documented in the inline doc comment above the call site.

File: iOS/DuckDuckGo/MainViewController.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift:1939]: Known and documented — see the doc comment on `loadUrlInNewTab` (lines 1920–1922): _"Timing relative to UI refresh differs by exit path: the URL-reuse branch fires this before `refreshOmniBar`/tab-bar refresh; all other branches fire it after. Callers should not rely on chrome state being settled inside the closure."_ No live caller hits the URL-reuse branch with a completion today.

File: iOS/DuckDuckGo/MainViewController.swift [hunk 3]
```diff
@@ -3422,9 +3437,50 @@ class MainViewController: UIViewController {
...
+        // crossing the Duck.ai ↔ web boundary. Chat → chat (e.g. browser app menu's "New Chat"
+        // tapped while already on a Duck.ai tab) stays in place — same rule the rest of the
+        // codebase enforces via `AIBoundaryNavigationDecision`. NTP transformations stay in
+        // place via the existing `link != nil` check. Gated on unified input so legacy chrome
+        // keeps its current in-tab behavior, except for deep links which already opened new
+        // tabs over existing content.
+        if unifiedToggleInputFeature.isAvailable || fromDeepLink,
+           let currentTab,
+           currentTab.tabModel.link != nil,
+           currentTab.isAITab == false {
```
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift]: Fixed in commit bc20d248e8 — the predicate now lets `fromDeepLink` qualify a content tab for the new-tab branch regardless of `isAITab`. App-icon long-press / widget / Universal Link from a Duck.ai tab with content now spawns a new chat tab, matching the doc-comment intent. UTI-driven boundary crosses still require `isAITab == false` so chat → chat from non-deep-link entry points stays in place.

File: iOS/DuckDuckGo/UnifiedToggleInput/MainViewController+UnifiedToggleInput.swift [hunk 4]
```diff
@@ -1230,32 +1219,47 @@ extension MainViewController: AIChatTabChatHeaderViewDelegate {
...
+    /// Chat content remains recoverable via Duck.ai → Recent chats.
+    func aiChatTabChatHeaderDidTapClose() {
+        guard let tab = currentTab?.tabModel else {
+            // The X is only visible from a Duck.ai tab header, which is only attached when a
+            // current tab exists — nil here means the header outlived its tab. Surface it.
+            assertionFailure("aiChatTabChatHeaderDidTapClose fired with no currentTab")
+            newTab(reuseExisting: true, allowingKeyboard: true)
+            return
+        }
+        closeTab(tab, behavior: .onlyClose)
```
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/UnifiedToggleInput/MainViewController+UnifiedToggleInput.swift:1260]: Stale — this review fired against commit 37a59c1e9f before the PR body was rewritten. The new test plan's step C5 documents `.onlyClose` and the tab-switcher selection rule (previous tab / next when current was first; fresh NTP only when the chat was the only tab) as the intended behavior. The old step 21 ("Land on NTP") no longer exists.

File: iOS/DuckDuckGo/MainViewController.swift [hunk 5]
```diff
@@ -3376,8 +3386,13 @@ class MainViewController: UIViewController {
...
         unifiedToggleInputCoordinator?.dismissOmnibarKeyboard()
 
+        // Open in a new tab when the current tab has real content AND we're actually crossing
+        // the Duck.ai ↔ web boundary. Chat → voice-chat (already on a Duck.ai tab) stays in
+        // place, mirroring the text-chat rule in `openAIChatInTab`. NTP transformations stay
+        // in place via the `link != nil` check.
         let hasContent = currentTab.tabModel.link != nil
-        let openInNewTab = hasContent && (unifiedToggleInputFeature.isAvailable || fromDeepLink)
+        let isCrossingBoundary = currentTab.isAITab == false
+        let openInNewTab = hasContent && isCrossingBoundary && (unifiedToggleInputFeature.isAvailable || fromDeepLink)
```
[the remark this answers is not in the document]
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift]: Stale — fixed in commit c7f2d1ed79 ("#5 voice always opens new tab from Plus — removed `isCrossingBoundary` from `openAIChatVoiceModeInTab` predicate"). Current code at MainViewController.swift:3397 reads `hasContent && (unifiedToggleInputFeature.isAvailable || fromDeepLink)` so a voice-mode deep link arriving on a Duck.ai tab DOES qualify for the new-tab branch, matching the text-chat `openAIChatInTab` rule.

File: iOS/DuckDuckGo/MainViewController.swift [hunk 6]
```diff
@@ -3422,9 +3433,39 @@ class MainViewController: UIViewController {
...
 
-        if fromDeepLink, let currentTab, currentTab.tabModel.link != nil {
-            let chatURL = currentTab.aiChatContentHandler.buildQueryURL(query: query, autoSend: autoSend, flowType: .default, tools: tools)
-            loadUrlInNewTab(chatURL, inheritedAttribution: nil)
+        // New tab when current has content: deep links always cross; UTI crosses only on web→chat (chat→chat from non-deep-link stays in-place, same rule as `AIBoundaryNavigationDecision`).
+        // NTP stays in-place via `link != nil`.
+        let shouldOpenInNewTab: Bool = {
+            guard let currentTab, currentTab.tabModel.link != nil else { return false }
+            if fromDeepLink { return true }
+            return unifiedToggleInputFeature.isAvailable && currentTab.isAITab == false
```
Comment (jaceklyp) [inline iOS/DuckDuckGo/MainViewController.swift]: **Minor:** the new-tab decision is reimplemented inline here instead of routed through `AIBoundaryNavigationDecision`, which was extracted to be the single source of truth for this matrix. I verified it *is* equivalent today — target is always AI in this path, so web→chat opens a new tab, chat→chat stays in place, with the `fromDeepLink`/`link != nil` overrides layered on — so this isn't a bug. But it's a divergence risk: a future change to the matrix in `AIBoundaryNavigationDecision` won't be reflected here.

Consider folding this into the enum so the rule lives in one place:

```swift
// current
let shouldOpenInNewTab: Bool = {
    guard let currentTab, currentTab.tabModel.link != nil else { return false }
    if fromDeepLink { return true }
    return unifiedToggleInputFeature.isAvailable && currentTab.isAITab == false
}()

// suggested
let shouldOpenInNewTab: Bool = {
    guard let currentTab, currentTab.tabModel.link != nil else { return false }
    if fromDeepLink { return true }
    return AIBoundaryNavigationDecision.forProgrammaticNavigation(
        currentIsAI: currentTab.isAITab,
        currentHasContent: true,
        targetIsAI: true,
        unifiedToggleInputAvailable: unifiedToggleInputFeature.isAvailable
    ) == .openInNewTab
}()
```

(Adjust arg labels to the actual signature.) Not blocking.

**Review by jaceklyp — COMMENTED** [review]
One minor maintainability note on the boundary rule — non-blocking.

**Comment by jaceklyp** [comment]
First batch of findings:

1. New Fire Tab  option in the menu should be behind fireMode feature flag; (I see it despite this flag being off for me and when I tap it opens as a regular tab)
2. Did we consider that fire button not only removes the chat but also dismisses the tab so we start on a NTP?
3. When user opens chats that X button in the upper left corner is extremely confusing then, people may think it closes that menu, but instead it closes the whole tab. Perhaps we should hide the button when menu is shown?
4. Bug: I was on a regular chat (I had UTI active, keyboard up); I tapped on + button, picked voice chat, it opened, but then I quickly closed it without saying anything, it dismissed and I showed regular chat again - keybord was down, but my UTI looked like it was focused (no app settings or fire button visible).
5. Glitch: I have 0 tabs opened, I see NTP, I press duck.ai button on the omnibar - duck.ai shows and the right pill on the nav bar flashes dark for a brief moment.
6. Bug? I'm on a duck.ai tab, I press + button and select New Chat, shouldn't we spawn a new tab? We don't.

**Comment by jaceklyp** [comment]
7. Bug - be on duck.ai site, tap app menu and select print - it prints some garbage ;) we should either fix or remove that option;

File: iOS/DuckDuckGo/MainViewController.swift (hunk 6 above)
Comment (aataraxiaa) [reply iOS/DuckDuckGo/MainViewController.swift]: Making the change.

**Review by jaceklyp — APPROVED** [review]

**Comment by aataraxiaa** [comment]
> First batch of findings:
> 
> 1. New Fire Tab  option in the menu should be behind fireMode feature flag; (I see it despite this flag being off for me and when I tap it opens as a regular tab)
✅ Fixed (disabling FF requires app restart)

> 2. Did we consider that fire button not only removes the chat but also dismisses the tab so we start on a NTP?
❌ We did on previous SR (for fire mode maybe..?). Decision was burn chat and remain on chat.

> 3. When user opens chats that X button in the upper left corner is extremely confusing then, people may think it closes that menu, but instead it closes the whole tab. Perhaps we should hide the button when menu is shown?
❌ Don’t find it extremely confusing as the user likely got there from the ’chats sidebar’ button which is beside the X and so the X was already visible. But if you feel strongly can you mention it in the SR Design subtask, get Bryan’s input.

> 4. Bug: I was on a regular chat (I had UTI active, keyboard up); I tapped on + button, picked voice chat, it opened, but then I quickly closed it without saying anything, it dismissed and I showed regular chat again - keybord was down, but my UTI looked like it was focused (no app settings or fire button visible).
✅ Fixed

> 5. Glitch: I have 0 tabs opened, I see NTP, I press duck.ai button on the omnibar - duck.ai shows and the right pill on the nav bar flashes dark for a brief moment.
❌ Can’t reproduce

> 6. Bug? I'm on a duck.ai tab, I press + button and select New Chat, shouldn't we spawn a new tab? We don’t.
❌ No, non-boundary (chat->chat) remains in the same tab.

**Comment by aataraxiaa** [comment]
> 7. Bug - be on duck.ai site, tap app menu and select print - it prints some garbage ;) we should either fix or remove that option;
❌ If I select print, I see the same thing that I see in production today, which is just like an image of the chat screen.
